aptos-move/framework/aptos-stdlib/sources/cryptography/groups.move

```diff
@@ -0,0 +1,749 @@
+    /// `Fq6` is an extension field of `Fq2`, constructed as `Fq6=Fq2[v]/(v^2-u-1)`.
+    /// `Fq12` is an extension field of `Fq6`, constructed as `Fq12=Fq6[w]/(w^2-v)`.
+    /// `BLS12_381_Gt` is a multiplicative subgroup of `Fq12`.
+    /// The order `r` of `BLS12_381_Gt` is 0x73eda753299d7d483339d80809a1d80553bda402fffe5bfeffffffff00000001. (Same as `BLS12_381_G1` and `BLS12_381_G2`.)
+    /// The identity of `BLS12_381_G2` is 1.
+    /// There exists a bilinear mapping from `(BLS12_381_G1, BLS12_381_G2)` to `BLS12_381_Gt`.
+    ///
+    /// A `Scalar<BLS12_381_G2>` is an integer between 0 and `r-1`.
+    ///
+    /// Functions `deserialize_scalar<BLS12_381_Gt>` and `serialize_scalar<BLS12_381_Gt>`
+    /// assume a 32-byte little-endian encoding of a `Scalar<BLS12_381_Gt>`.
+    ///
+    /// An `Element<BLS12_381_Gt>` is an element in group `BLS12_381_Gt`.
+    ///
+    /// Functions `serialize_element_uncompressed<BLS12_381_Gt>` and `deserialize_element_uncompressed<BLS12_381_Gt>`,
+    /// as well as `serialize_element_ompressed<BLS12_381_Gt>` and `deserialize_element_compressed<BLS12_381_Gt>`,
+    /// assume a 576-byte encoding `[b_0, ..., b_575]` of an `Element<BLS12_381_Gt>`, with the following rules.
+    ///     - Assume the given element is `e=c_0+c_1*w` where `c_i=c_i0+c_i1*v+c_i2*v^2 for i=0..1` and `c_ij=c_ij0+c_ij1*u for j=0..2`.
+    ///     - `[b_0, ..., b_575]` is a concatenation of 12 encoded `Fq` elements: `c_000, c_001, c_010, c_011, c_020, c_021, c_100, c_101, c_110, c_111, c_120, c_121`.
+    ///     - Every `c_ijk` uses a 48-byte little-endian encoding.
+    struct BLS12_381_Gt {}
+
+    /// This struct represents an integer between 0 and `r-1`, where `r` is the order of group `G`.
+    struct Scalar<phantom G> has copy, drop {
+        handle: u64
+    }
+
+    /// This struct represents an element of the group represented by the type argument `G`.
+    struct Element<phantom G> has copy, drop {
+        handle: u64
+    }
+
+    /// Perform a pairing.
+    public fun pairing<G1,G2,Gt>(element_1: &Element<G1>, element_2: &Element<G2>): Element<Gt> {
+        abort_if_feature_disabled();
+        Element<Gt> {
+            handle: pairing_product_internal<G1,G2,Gt>(vector[element_1.handle], vector[element_2.handle])
+        }
+    }
+
+    /// Compute the product of multiple pairing.
+    public fun pairing_product<G1, G2, Gt>(g1_elements: &vector<Element<G1>>, g2_elements: &vector<Element<G2>>): Element<Gt> {
+        abort_if_feature_disabled();
+        let num_g1 = std::vector::length(g1_elements);
+        let num_g2 = std::vector::length(g2_elements);
+        assert!(num_g1 == num_g2, std::error::invalid_argument(1));
+        let g1_handles = vector[];
+        let g2_handles = vector[];
+        let i = 0;
+        while (i < num_g2) {
+            std::vector::push_back(&mut g1_handles, std::vector::borrow(g1_elements, i).handle);
```

Comment:
You can also do this parsing inside the native rather than here; see:

- `multi_scalar_mul` native [here](https://github.com/aptos-labs/aptos-core/blob/21991ffa96acde5ec211ba73dea42528a29b1e01/aptos-move/framework/aptos-stdlib/sources/cryptography/ristretto255.move#L275)
- native implementation [here](https://github.com/aptos-labs/aptos-core/blob/9826f0304fa523de16eb12bf837a4be46c684764/aptos-move/framework/src/natives/cryptography/ristretto255_point.rs#L511)

Reply:
But might be more involved due to the type argument. Nevermind. Marking as resolved.